src/bootstrap/src/core/build_steps/doc.rs

```diff
@@ -20,10 +20,29 @@ use crate::core::builder::{
 use crate::core::config::{Config, TargetSelection};
 use crate::helpers::{is_path_in_submodule, symlink_dir, t, up_to_date};
 
+/// If the doc path is inside a submodule and is thus not the same as the submodule path, then we
+/// need to use the submodule path to checkout the submodule. The doc could be inside an submodule
+/// that is not checked out. Helper for [`book`].
+fn select_book_submodule_path<'a>(doc_path: &'a str, submodule_path: Option<&'a str>) -> &'a str {
+    submodule_path.unwrap_or(doc_path)
+}
+
 macro_rules! book {
-    ($($name:ident, $path:expr, $book_name:expr, $lang:expr ;)+) => {
-        $(
-            #[derive(Debug, Clone, Hash, PartialEq, Eq)]
+    (
+        $name:ident {
+            name: $book_name:expr,
+            // This can be *within* a submodule, e.g. `src/tools/cargo/src/doc`, compared to the
+            // submodule *itself*.
+            doc_path: $doc_path:expr
+            // Optional; only needed if `doc_path` is inside the submodule path, i.e. they are not
+            // the same.
+            $(, submodule_path: $submodule_path:expr )?
+            // Optional; only needed if the book has multi-lingual support.
+            $(, languages: $languages:expr )?
+            $( , )?
+        }
+    ) => {
+        #[derive(Debug, Clone, Hash, PartialEq, Eq)]
```

Comment:
Please correct me if I am wrong: the problem is that we are trying to update a submodule that is actually not a submodule but a subdirectory within a submodule. If that's the case, there is no need to overcomplicate this macro again (which kills the purpose of https://github.com/rust-lang/rust/pull/134967). Instead, make [is_path_in_submodule](https://github.com/onur-ozkan/rust/blob/6eb9ebf1d61131df43f18e379493e23a193c3a7c/src/bootstrap/src/utils/helpers.rs#L63-L64) to return the submodule path as `Some(path)` instead of returning `true` and return `None` instead of `false`.

Reply:
Sounds reasonable, let me do that a bit later.

Reply:
Changed `is_path_in_submodule` to `submodule_path_of`, which returns `Some(submodule_path)` if the destination is inside a submodule, and return `None` if it's not a submodule we handle.

Reply:
Thanks